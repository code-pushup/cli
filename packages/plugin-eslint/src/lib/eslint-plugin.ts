import { mkdir, writeFile } from 'node:fs/promises';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { PluginConfig } from '@code-pushup/models';
import { name, version } from '../../package.json';
import { ESLintPluginConfig, eslintPluginConfigSchema } from './config';
import { listAuditsAndGroups } from './meta';
import { ESLINTRC_PATH, createRunnerConfig } from './runner';
import { setupESLint } from './setup';

/**
 * Instantiates Code PushUp ESLint plugin for use in core config.
 *
 * @example
 * import eslintPlugin from '@code-pushup/eslint-plugin'
 *
 * export default {
 *   // ... core config ...
 *   plugins: [
 *     // ... other plugins ...
 *     await eslintPlugin({
 *       eslintrc: '.eslintrc.json',
 *       patterns: ['src', 'test/*.spec.js']
 *     })
 *   ]
 * }
 *
 * @param config Configuration options.
 * @returns Plugin configuration as a promise.
 */
export async function eslintPlugin(
  config: ESLintPluginConfig,
): Promise<PluginConfig> {
  const { eslintrc, patterns } = eslintPluginConfigSchema.parse(config);

  const eslint = setupESLint(eslintrc);

  const { audits, groups } = await listAuditsAndGroups(eslint, patterns);

  // save inline config to file so runner can access it later
  if (typeof eslintrc !== 'string') {
    await mkdir(dirname(ESLINTRC_PATH), { recursive: true });
    await writeFile(ESLINTRC_PATH, JSON.stringify(eslintrc));
  }
  const eslintrcPath = typeof eslintrc === 'string' ? eslintrc : ESLINTRC_PATH;

  const runnerScriptPath = join(
    fileURLToPath(dirname(import.meta.url)),
    'bin.js',
  );

  return {
    slug: 'eslint',
    title: 'ESLint',
    icon: 'eslint',
    description: 'Official Code PushUp ESLint plugin',
    docsUrl: 'https://www.npmjs.com/package/@code-pushup/eslint-plugin',
    packageName: name,
    version,

    audits,
    groups,

    runner: await createRunnerConfig(
      runnerScriptPath,
      audits,
      eslintrcPath,
      patterns,
    ),
  };
}
