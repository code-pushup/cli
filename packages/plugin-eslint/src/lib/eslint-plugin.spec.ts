import { dirname, join } from 'path';
import { fileURLToPath } from 'url';
import type { SpyInstance } from 'vitest';
import { eslintPlugin } from './eslint-plugin';

describe('eslintPlugin', () => {
  const fixturesDir = join(
    fileURLToPath(dirname(import.meta.url)),
    '..',
    '..',
    'test',
    'fixtures',
  );

  let cwdSpy: SpyInstance;

  beforeAll(() => {
    cwdSpy = vi.spyOn(process, 'cwd');
  });

  afterAll(() => {
    cwdSpy.mockRestore();
  });

  it('should initialize ESLint plugin for React application', async () => {
    cwdSpy.mockReturnValue(join(fixturesDir, 'todos-app'));
    await expect(
      eslintPlugin({
        eslintrc: '.eslintrc.js',
        patterns: ['src/**/*.js', 'src/**/*.jsx'],
      }),
    ).resolves.toMatchSnapshot();
  });

  it('should initialize ESLint plugin for Nx project', async () => {
    cwdSpy.mockReturnValue(join(fixturesDir, 'nx-monorepo'));
    await expect(
      eslintPlugin({
        eslintrc: './packages/utils/.eslintrc.json',
        patterns: ['packages/utils/**/*.ts', 'packages/utils/**/*.json'],
      }),
    ).resolves.toMatchSnapshot();
  });

  it('should throw when invalid parameters provided', async () => {
    await expect(
      // @ts-expect-error simulating invalid non-TS config
      eslintPlugin({ eslintrc: '.eslintrc.json' }),
    ).rejects.toThrowError('patterns');
  });

  it("should throw if eslintrc file doesn't exist", async () => {
    await expect(
      eslintPlugin({ eslintrc: '.eslintrc.yml', patterns: '**/*.js' }),
    ).rejects.toThrowError(
      'Failed to load config ".eslintrc.yml" to extend from',
    );
  });
});
