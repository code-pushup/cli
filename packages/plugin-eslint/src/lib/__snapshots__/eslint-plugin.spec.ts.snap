// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`eslintPlugin > should initialize ESLint plugin 1`] = `
{
  "audits": [
    {
      "description": "ESLint rule **no-cond-assign**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/no-cond-assign",
      "slug": "no-cond-assign",
      "title": "Disallow assignment operators in conditional expressions",
    },
    {
      "description": "ESLint rule **no-const-assign**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/no-const-assign",
      "slug": "no-const-assign",
      "title": "Disallow reassigning \`const\` variables",
    },
    {
      "description": "ESLint rule **no-debugger**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/no-debugger",
      "slug": "no-debugger",
      "title": "Disallow the use of \`debugger\`",
    },
    {
      "description": "ESLint rule **no-invalid-regexp**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/no-invalid-regexp",
      "slug": "no-invalid-regexp",
      "title": "Disallow invalid regular expression strings in \`RegExp\` constructors",
    },
    {
      "description": "ESLint rule **no-undef**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/no-undef",
      "slug": "no-undef",
      "title": "Disallow the use of undeclared variables unless mentioned in \`/*global */\` comments",
    },
    {
      "description": "ESLint rule **no-unreachable-loop**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/no-unreachable-loop",
      "slug": "no-unreachable-loop",
      "title": "Disallow loops with a body that allows only one iteration",
    },
    {
      "description": "ESLint rule **no-unsafe-negation**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/no-unsafe-negation",
      "slug": "no-unsafe-negation",
      "title": "Disallow negating the left operand of relational operators",
    },
    {
      "description": "ESLint rule **no-unsafe-optional-chaining**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/no-unsafe-optional-chaining",
      "slug": "no-unsafe-optional-chaining",
      "title": "Disallow use of optional chaining in contexts where the \`undefined\` value is not allowed",
    },
    {
      "description": "ESLint rule **no-unused-vars**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/no-unused-vars",
      "slug": "no-unused-vars",
      "title": "Disallow unused variables",
    },
    {
      "description": "ESLint rule **use-isnan**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/use-isnan",
      "slug": "use-isnan",
      "title": "Require calls to \`isNaN()\` when checking for \`NaN\`",
    },
    {
      "description": "ESLint rule **valid-typeof**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/valid-typeof",
      "slug": "valid-typeof",
      "title": "Enforce comparing \`typeof\` expressions against valid strings",
    },
    {
      "description": "ESLint rule **arrow-body-style**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/arrow-body-style",
      "slug": "arrow-body-style",
      "title": "Require braces around arrow function bodies",
    },
    {
      "description": "ESLint rule **camelcase**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/camelcase",
      "slug": "camelcase",
      "title": "Enforce camelcase naming convention",
    },
    {
      "description": "ESLint rule **curly**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/curly",
      "slug": "curly",
      "title": "Enforce consistent brace style for all control statements",
    },
    {
      "description": "ESLint rule **eqeqeq**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/eqeqeq",
      "slug": "eqeqeq",
      "title": "Require the use of \`===\` and \`!==\`",
    },
    {
      "description": "ESLint rule **max-lines-per-function**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/max-lines-per-function",
      "slug": "max-lines-per-function",
      "title": "Enforce a maximum number of lines of code in a function",
    },
    {
      "description": "ESLint rule **max-lines**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/max-lines",
      "slug": "max-lines",
      "title": "Enforce a maximum number of lines per file",
    },
    {
      "description": "ESLint rule **no-shadow**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/no-shadow",
      "slug": "no-shadow",
      "title": "Disallow variable declarations from shadowing variables declared in the outer scope",
    },
    {
      "description": "ESLint rule **no-var**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/no-var",
      "slug": "no-var",
      "title": "Require \`let\` or \`const\` instead of \`var\`",
    },
    {
      "description": "ESLint rule **object-shorthand**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/object-shorthand",
      "slug": "object-shorthand",
      "title": "Require or disallow method and property shorthand syntax for object literals",
    },
    {
      "description": "ESLint rule **prefer-arrow-callback**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/prefer-arrow-callback",
      "slug": "prefer-arrow-callback",
      "title": "Require using arrow functions for callbacks",
    },
    {
      "description": "ESLint rule **prefer-const**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/prefer-const",
      "slug": "prefer-const",
      "title": "Require \`const\` declarations for variables that are never reassigned after declared",
    },
    {
      "description": "ESLint rule **prefer-object-spread**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/prefer-object-spread",
      "slug": "prefer-object-spread",
      "title": "Disallow using Object.assign with an object literal as the first argument and prefer the use of object spread instead",
    },
    {
      "description": "ESLint rule **yoda**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/yoda",
      "slug": "yoda",
      "title": "Require or disallow \\"Yoda\\" conditions",
    },
    {
      "description": "ESLint rule **jsx-key**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/jsx-key.md",
      "slug": "react-jsx-key",
      "title": "Disallow missing \`key\` props in iterators/collection literals",
    },
    {
      "description": "ESLint rule **prop-types**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/prop-types.md",
      "slug": "react-prop-types",
      "title": "Disallow missing props validation in a React component definition",
    },
    {
      "description": "ESLint rule **react-in-jsx-scope**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/react-in-jsx-scope.md",
      "slug": "react-react-in-jsx-scope",
      "title": "Disallow missing React when using JSX",
    },
    {
      "description": "ESLint rule **rules-of-hooks**, from _react-hooks_ plugin.",
      "docsUrl": "https://reactjs.org/docs/hooks-rules.html",
      "slug": "react-hooks-rules-of-hooks",
      "title": "enforces the Rules of Hooks",
    },
    {
      "description": "ESLint rule **exhaustive-deps**, from _react-hooks_ plugin.",
      "docsUrl": "https://github.com/facebook/react/issues/14920",
      "slug": "react-hooks-exhaustive-deps",
      "title": "verifies the list of dependencies for Hooks like useEffect and similar",
    },
    {
      "description": "ESLint rule **display-name**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/display-name.md",
      "slug": "react-display-name",
      "title": "Disallow missing displayName in a React component definition",
    },
    {
      "description": "ESLint rule **jsx-no-comment-textnodes**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/jsx-no-comment-textnodes.md",
      "slug": "react-jsx-no-comment-textnodes",
      "title": "Disallow comments from being inserted as text nodes",
    },
    {
      "description": "ESLint rule **jsx-no-duplicate-props**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/jsx-no-duplicate-props.md",
      "slug": "react-jsx-no-duplicate-props",
      "title": "Disallow duplicate properties in JSX",
    },
    {
      "description": "ESLint rule **jsx-no-target-blank**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/jsx-no-target-blank.md",
      "slug": "react-jsx-no-target-blank",
      "title": "Disallow \`target=\\"_blank\\"\` attribute without \`rel=\\"noreferrer\\"\`",
    },
    {
      "description": "ESLint rule **jsx-no-undef**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/jsx-no-undef.md",
      "slug": "react-jsx-no-undef",
      "title": "Disallow undeclared variables in JSX",
    },
    {
      "description": "ESLint rule **jsx-uses-react**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/jsx-uses-react.md",
      "slug": "react-jsx-uses-react",
      "title": "Disallow React to be incorrectly marked as unused",
    },
    {
      "description": "ESLint rule **jsx-uses-vars**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/jsx-uses-vars.md",
      "slug": "react-jsx-uses-vars",
      "title": "Disallow variables used in JSX to be incorrectly marked as unused",
    },
    {
      "description": "ESLint rule **no-children-prop**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/no-children-prop.md",
      "slug": "react-no-children-prop",
      "title": "Disallow passing of children as props",
    },
    {
      "description": "ESLint rule **no-danger-with-children**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/no-danger-with-children.md",
      "slug": "react-no-danger-with-children",
      "title": "Disallow when a DOM element is using both children and dangerouslySetInnerHTML",
    },
    {
      "description": "ESLint rule **no-deprecated**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/no-deprecated.md",
      "slug": "react-no-deprecated",
      "title": "Disallow usage of deprecated methods",
    },
    {
      "description": "ESLint rule **no-direct-mutation-state**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/no-direct-mutation-state.md",
      "slug": "react-no-direct-mutation-state",
      "title": "Disallow direct mutation of this.state",
    },
    {
      "description": "ESLint rule **no-find-dom-node**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/no-find-dom-node.md",
      "slug": "react-no-find-dom-node",
      "title": "Disallow usage of findDOMNode",
    },
    {
      "description": "ESLint rule **no-is-mounted**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/no-is-mounted.md",
      "slug": "react-no-is-mounted",
      "title": "Disallow usage of isMounted",
    },
    {
      "description": "ESLint rule **no-render-return-value**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/no-render-return-value.md",
      "slug": "react-no-render-return-value",
      "title": "Disallow usage of the return value of ReactDOM.render",
    },
    {
      "description": "ESLint rule **no-string-refs**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/no-string-refs.md",
      "slug": "react-no-string-refs",
      "title": "Disallow using string references",
    },
    {
      "description": "ESLint rule **no-unescaped-entities**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/no-unescaped-entities.md",
      "slug": "react-no-unescaped-entities",
      "title": "Disallow unescaped HTML entities from appearing in markup",
    },
    {
      "description": "ESLint rule **no-unknown-property**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/no-unknown-property.md",
      "slug": "react-no-unknown-property",
      "title": "Disallow usage of unknown DOM property",
    },
    {
      "description": "ESLint rule **no-unsafe**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/no-unsafe.md",
      "slug": "react-no-unsafe",
      "title": "Disallow usage of unsafe lifecycle methods",
    },
    {
      "description": "ESLint rule **require-render-return**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/require-render-return.md",
      "slug": "react-require-render-return",
      "title": "Enforce ES5 or ES6 class for returning value in render function",
    },
  ],
  "description": "Official Code PushUp ESLint plugin",
  "icon": "eslint",
  "runner": {
    "args": [
      "eslint",
      "--config=.eslintrc.js",
      "--format=json",
      "--output-file=tmp/out.json",
      "src/**/*.js",
      "src/**/*.jsx",
    ],
    "command": "npx",
    "outputPath": "tmp/out.json",
  },
  "slug": "eslint",
  "title": "ESLint",
}
`;
